ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java

```diff
@@ -0,0 +1,294 @@
+    try {
+      createMySqlAccountStore();
+    } catch (SQLException e) {
+      logger.error("MySQL account store creation failed", e);
+      // Continue account service creation. Cache will initialized with metadata from backup copy on local disk to serve read requests.
+      // Write requests will be blocked until MySql DB is up. Connection to MySql DB will be retried in polling thread that fetches new accounts.
+    }
+
+    // TODO: create backup manager to manage local back up copies of Account and container metadata and lastModifiedTime
+
+    initializeCache();
+
+    // TODO: Start background thread for periodically querying MYSql DB for added/modified accounts and containers. Also, retry creation of MySqlAccountStore if it failed above.
+
+    // TODO: Subscribe to notifications from ZK
+  }
+
+  /**
+   * creates MySql Account store which establishes connection to database
+   * @throws SQLException
+   */
+  private void createMySqlAccountStore() throws SQLException {
+    if (mySqlAccountStore == null) {
+      try {
+        mySqlAccountStore = new MySqlAccountStore(this.mySqlConfig);
+      } catch (SQLException e) {
+        logger.error("MySQL account store creation failed", e);
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Call to initialize in-memory cache by fetching all the {@link Account}s and {@link Container}s metadata records.
+   * It consists of 2 steps:
+   * 1. Check local disk for back up copy and load metadata and last modified time of Accounts/Containers into cache.
+   * 2. Fetch added/modified accounts and containers from mysql database since the last modified time (found in step 1)
+   *    and load into cache.
+   */
+  void initializeCache() {
+    // TODO: Check local disk for back up copy and load metadata and last modified time into cache.
+    fetchAndUpdateCache();
+  }
+
+  /**
+   * Fetches all the accounts and containers that have been created or modified since the last sync time and loads into
+   * cache.
+   */
+  void fetchAndUpdateCache() {
+    try {
+      // Retry connection to mysql if we couldn't set up previously
+      createMySqlAccountStore();
+    } catch (SQLException e) {
+      logger.error("Fetching Accounts from MySql DB failed", e);
+      return;
+    }
+
+    // get the last sync time of accounts and containers in cache
+    long lastModifiedTime = accountInfoMap.getLastModifiedTime();
+
+    try {
+      // Fetch all added/modified accounts from MySql database since LMT
+      List<Account> accounts = mySqlAccountStore.getNewAccounts(lastModifiedTime);
+      rwLock.writeLock().lock();
+      try {
+        accountInfoMap.updateAccounts(accounts);
+      } finally {
+        rwLock.writeLock().unlock();
+      }
+
+      // Fetch all added/modified containers from MySql database since LMT
+      List<Container> containers = mySqlAccountStore.getNewContainers(lastModifiedTime);
+      rwLock.writeLock().lock();
```

Comment:
I would recommend holding the lock for both account and container update, to make sure consumers see a consistent view of the cache.

Reply:
Sure, yeah true. Made changes to hold the lock for both account and container updates.